sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,518 @@
+      extends UnboundedSource<byte[], UnboundedSource.CheckpointMark> {
+
+    private Read spec;
+
+    public UnboundedMqttSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
+                                        CheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      List<UnboundedMqttSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        // NB: it's important that user understand the impact of MQTT message QoS
+        sources.add(new UnboundedMqttSource(spec));
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return VoidCoder.of();
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+  }
+
+  /**
+   * POJO used to store MQTT message and its timestamp.
+   */
+  private static class MqttMessageWithTimestamp {
+
+    private MqttMessage message;
+    private Instant timestamp;
+
+    public MqttMessageWithTimestamp(MqttMessage message, Instant timestamp) {
+      this.message = message;
+      this.timestamp = timestamp;
+    }
+
+    public MqttMessage getMessage() {
+      return message;
+    }
+
+    public void setMessage(MqttMessage message) {
+      this.message = message;
+    }
+
+    public Instant getTimestamp() {
+      return timestamp;
+    }
+
+    public void setTimestamp(Instant timestamp) {
+      this.timestamp = timestamp;
+    }
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader<byte[]> {
+
+    private final UnboundedMqttSource source;
+
+    private MqttClient client;
+    private byte[] current;
+    private Instant currentTimestamp;
+    private BlockingQueue<MqttMessageWithTimestamp> queue;
+
+    private UnboundedMqttReader(UnboundedMqttSource source) {
+      this.source = source;
+      this.current = null;
+      this.queue = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOGGER.debug("Starting MQTT reader");
+      Read spec = source.spec;
+      try {
+        client = spec.mqttConnectionConfiguration().getClient();
+        client.subscribe(spec.mqttConnectionConfiguration().topic());
+        client.setCallback(new MqttCallback() {
+          @Override
+          public void connectionLost(Throwable cause) {
+            LOGGER.warn("MQTT connection lost", cause);
+            try {
+              close();
```

Comment:
Actually, the call to `close()` doesn't make sense. My purpose was to "force" the client disconnect and close. But it's not a good idea: the connection can be lost but not definitely, it can come back later. So, I should not close the client when the connection is lost. I think it's safer.